### ISSUE TITLE
Various test fails

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#551]
I am trying to build BOUT++ on fedora rawhide.
There are several issues:

On `x86_64` only `drift-instability` fails:
```
Making resistive drift instability test
Running drift instability test, zeff =  2
Traceback (most recent call last):
  File "./runtest", line 111, in <module>
    lbNorm = lZeta*((grid['Bpxy'])[0,old_div(ny,2)] / (grid['Bxy'])[0,old_div(ny,2)])  # binormal coord range [cm]
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
rm: cannot remove 'data_1/BOUT.dmp.*': No such file or directory
rm: cannot remove 'data_10/BOUT.dmp.*': No such file or directory
```

On `i686` in the `test-initial` the `mixmode` and `mixmode_seed` fail - probably a usage of float instead of double.

Why `arm` and some more exotic architectures fail: not sure.
But the binaries crash in all cases.

Full logs are available here:
https://koji.fedoraproject.org/koji/taskinfo?taskID=19383165

**Comment by bendudson** [comment]
Could this be a problem with old_div? The runtest script tries to import it from past.utils, but if it's not defined it uses
    def old_div(a,b):
      return a / b
https://github.com/boutproject/BOUT-dev/blob/master/examples/drift-instability/runtest#L19

Perhaps this is returning a float rather than an integer? Replacing this with
    def old_div(a,b):
      return int(a / b)
should fix it if this is the problem

**Comment by bendudson** [comment]
The errors on ppc (https://kojipkgs.fedoraproject.org//work/tasks/3167/19383167/build.log) are odd...

in test-field:

    Couldn't get integer from mesh:nx = '6'
    Mesh::get(ival) on line 59 of 'mesh.cxx'

which is reading this line: https://github.com/boutproject/BOUT-dev/blob/master/examples/test-field/data/BOUT.inp#L5

from here:
https://github.com/boutproject/BOUT-dev/blob/master/src/mesh/mesh.cxx#L58

through here:
https://github.com/boutproject/BOUT-dev/blob/master/src/mesh/data/gridfromoptions.cxx#L18

to the parser call here which returns null:
https://github.com/boutproject/BOUT-dev/blob/master/src/sys/options.cxx#L79

which checks a cache here:
https://github.com/boutproject/BOUT-dev/blob/master/src/field/field_factory.cxx#L337

creates a lexer to parse the string here:
https://github.com/boutproject/BOUT-dev/blob/master/src/sys/expressionparser.cxx#L160

which parses the string to get the first token (should be '6') here:
https://github.com/boutproject/BOUT-dev/blob/master/src/sys/expressionparser.cxx#L332

which is sent to be evaluated as an expression (an integer) here:
https://github.com/boutproject/BOUT-dev/blob/master/src/sys/expressionparser.cxx#L322

and finally the FieldValue object is created here:
https://github.com/boutproject/BOUT-dev/blob/master/src/sys/expressionparser.cxx#L248

and cached in a list for cleanup here:
https://github.com/boutproject/BOUT-dev/blob/master/include/bout/sys/expressionparser.hxx#L144

"All problems in computer science can be solved by another level of indirection, except for the problem of too many layers of indirection" -- David J Wheeler

**Comment by bendudson** [comment]
The arm one is quite interesting... 
https://kojipkgs.fedoraproject.org//work/tasks/3172/19383172/build.log

expressionparser.cxx:246:3: warning: case label value is less than minimum value for type
   case -1: { // a number
   ^~~~

which means that the options will never be able to read a number. The switch is on the variable lex.curtok :https://github.com/boutproject/BOUT-dev/blob/master/src/sys/expressionparser.cxx#L245

which is defined as type "char" here:
https://github.com/boutproject/BOUT-dev/blob/master/include/bout/sys/expressionparser.hxx#L127

It seems that GCC defines char as signed (which is what I had been expecting), but it's actually implementation defined:
http://stackoverflow.com/questions/2054939/is-char-signed-or-unsigned-by-default

For now easiest may be to make this "signed char", since negative values are used to indicate numbers and strings, and positive values other tokens like "+".

**Comment by dschwoerer** [comment]
Nice spot with the arm:
https://kojipkgs.fedoraproject.org//work/tasks/9471/19389471/build.log

Just as note: These builds are all done with gcc

Concerning the test-initial:
I confirmed that the C++ implementation is causing the issue, not the python.

I am not sure whether this is relevant, but in `x` in https://github.com/boutproject/BOUT-dev/blob/master/src/field/fieldgenerators.cxx#L237 is slightly different:
```
32 bit: 0.40651446929667556240 0.21954340789002668721 0.03257234648337781202 0.82965954118274165552 0.64268847977609278033 0.45571741836944390514 0.26874635696279502994 0.08177529555614616863 0.87886249025550999825 0.69189142884886112306 0.50492036744221224787 0.31794930603556337267 0.13097824462891449748 0.92806543932827834098
64 bit: 0.40651446929667556240 0.21954340789002668721 0.03257234648337781202 0.82965954118274165552 0.64268847977609278033 0.45571741836944390514 0.26874635696279502994 0.08177529555614615475 0.87886249025550999825 0.69189142884886112306 0.50492036744221224787 0.31794930603556337267 0.13097824462891452524 0.92806543932827834098
```
after the iteration, in https://github.com/boutproject/BOUT-dev/blob/master/src/field/fieldgenerators.cxx#L242 the difference is grown:
```
32 bit: 0.32073302593534186444 0.52237685016928614434 0.27543155608418895541 0.01268374869204529944 0.14670583459877434285 0.25697276313639239032 0.50908431593784497782 0.72014968777804322375 0.26064818865970734718 0.24049486565784328196 0.54396366434473730767 0.22166861848282992664 0.12641071774721521215 0.92394177705013613711
64 bit: 0.32073151393668136722 0.52237714853800687465 0.27542908431845519734 0.01268416530104616094 0.14669967425394928040 0.25698144278836088095 0.50907883774723095005 0.72013554959517112053 0.26070248842444454507 0.24049488872114049842 0.54396364240185679506 0.22201577822624740666 0.12346438977794303715 0.92111719820421689153
```
It is now big enough to explain the failure of the test
So why is there a difference in the first place?

```
32 bit: 0.4065144692966755624041752525954507291317 -> 0.3207330259353418644430178119364427402616 0.2195434078900266872125257577863521873951 -> 0.5223768501692861443430615508987102657557 0.0325723464833778120208762629772536456585 -> 0.2754315560841889554133388173795538023114 0.8296595411827416555183845048304647207260 -> 0.0126837486920452994360042353605422249530 0.6426884797760927803267350100213661789894 -> 0.1467058345987743428473493167984997853637 0.4557174183694439051350855152122676372528 -> 0.2569727631363923903151658123533707112074 0.2687463569627950299434360204031690955162 -> 0.5090843159378449778174058337754104286432 0.0817752955561461686295743334085273090750 -> 0.7201496877780432237514673943223897367716 0.8788624902555099982492947674472816288471 -> 0.2606481886597073471811825129407225176692 0.6918914288488611230576452726381830871105 -> 0.2404948656578432819586055302352178841829 0.5049203674422122478659957778290845453739 -> 0.5439636643447373076654116630379576236010 0.3179493060355633726743462830199860036373 -> 0.2216686184828299266413864643254783004522 0.1309782446289144974826967882108874619007 -> 0.1264107177472152121477932951165712438524 0.9280654393282783409802050300640985369682 -> 0.9239417770501361371060511373798362910748
64 bit: 0.4065144692966755624041752525954507291317 -> 0.3207315139366813672161526937998132780194 0.2195434078900266872125257577863521873951 -> 0.5223771485380068746451343031367287039757 0.0325723464833778120208762629772536456585 -> 0.2754290843184551973443774386396398767829 0.8296595411827416555183845048304647207260 -> 0.0126841653010461609396575255459538311698 0.6426884797760927803267350100213661789894 -> 0.1466996742539492804002776438210275955498 0.4557174183694439051350855152122676372528 -> 0.2569814427883608809466409184096846729517 0.2687463569627950299434360204031690955162 -> 0.5090788377472309500504366042150650173426 0.0817752955561461547517865255940705537796 -> 0.7201355495951711205293577222619205713272 0.8788624902555099982492947674472816288471 -> 0.2607024884244445450676153086533304303885 0.6918914288488611230576452726381830871105 -> 0.2404948887211404984221019276446895673871 0.5049203674422122478659957778290845453739 -> 0.5439636424018567950611213746014982461929 0.3179493060355633726743462830199860036373 -> 0.2220157782262474066570234754181001335382 0.1309782446289145252382724038398009724915 -> 0.1234643897779430371519637787969259079546 0.9280654393282783409802050300640985369682 -> 0.9211171982042168915327806644199881702662  
```

If we print 40 decimal digits, we should by far have all bits.
In some cases they are initially the same, and afterwards they arent.
The code is this:
```C++
  printf("%.40f -> ",x);
  // Iterate logistic map                                                                                                                                                         
  for(int i=0;i!=niter;++i)
    x = 3.99 * x * (1. - x);
  printf("%.40f ",x)
```
as far as I could find, fp numerical constants are always [double](http://www.cplusplus.com/doc/tutorial/constants/) - and taking into account how small the difference is, it doesn't seem this is an issue with single/double precision, but rather seems like a bug in rounding in the IEEE754 implementation. I compiled both with -O0 (at least src/field/*).

**Comment by dschwoerer** [comment]
i686 uses [x87](https://en.wikipedia.org/wiki/X87) for arithmetic - which is too precise.
```
       -ffloat-store
           Do not store floating-point variables in registers, and inhibit
           other options that might change whether a floating-point value is
           taken from a register or memory.

           This option prevents undesirable excess precision on machines such
           as the 68000 where the floating registers (of the 68881) keep more
           precision than a "double" is supposed to have.  Similarly for the
           x86 architecture.  For most programs, the excess precision does
           only good, but a few programs rely on the precise definition of
           IEEE floating point.  Use -ffloat-store for such programs, after
           modifying them to store all pertinent intermediate computations
           into variables.
```
See also: https://bugzilla.redhat.com/show_bug.cgi?id=1447804

**Comment by dschwoerer** [comment]
Fixed by #553 #554 and #555
